src/components/Menu.jsx

```diff
@@ -12,6 +12,8 @@ const propTypes = {
   renderItemLabel: PropTypes.func,
   value: PropTypes.string,
   onSelect: PropTypes.func,
+  shouldTransferFocus: PropTypes.bool,
+  notifyBeforeFocusWrap: PropTypes.func,
 };
```

Comment:
I've written something that looks almost identical to what you've described: 

`if (onBeforeItemFocusChange) {
    focusedIndex = onBeforeItemFocusChange(focusedIndex, nextFocusedIndex, event.key, options.length);
  // Otherwise, continue as usual
  } else {
    if (nextFocusedIndex >= options.length) {
            focusedIndex = 0;
    }
}`

And for AutocompleteInput: 

`handleQuickAddOnBeforeItemFocusChange(currentFocusedIndex, nextFocusedIndex, eventKey, itemsLength) {
    if (nextFocusedIndex >= itemsLength) {
      this.dropdownMenuInput.focusMenu(0);
      return null;
    } else if (nextFocusedIndex < 0) {
      this.dropdownMenuInput.focusMenu(-1);
      return null;
    } else {
      return nextFocusedIndex;
    }
  }`

And for QuickAdd:
`focusMenu(itemIndex) {
    if (this.menu) {
      this.menu.focus(itemIndex);
    }
  }`

And for Menu, something like...: 

`focus(itemIndex) {

    let {
      focusedIndex,
    } = this.state;

    const { 
      options,
    } = this.props;

    focusedIndex = itemIndex === -1 ? options.length -1: itemIndex;
    
    this.setState({
      focusedIndex,
    });

    if (this.domNode) {
      this.domNode.focus();
    }
  }`

Although this works somewhat, the problem I see is with Menu's focus(itemIndex) method. I still need to work out some of the logic and ensure that it remains mostly presentational, or 'dumb', like you said. Overall, I think it's a much better implementation than the initial pull request. Initial thoughts? I'll have this up later tonight or early tomorrow morning for you to look over in its entirety.